go/kbfs/libkbfs/folder_branch_ops.go

```diff
@@ -8503,6 +8501,7 @@ func (fbo *folderBranchOps) Reset(
 		return err
 	}
 	oldHandle.SetFinalizedInfo(finalizedInfo)
+	// This can't be susceptible to the WaitGroup due to a potential deadlock.
```

Comment:
I think it might have been a similar situation - observers being evaluated at the wrong time.

Reply:
Nope,
```
goroutine 24259 [semacquire]:
sync.runtime_SemacquireMutex(0xc013b85e64, 0x4041a00)
        /Users/jzila/.gvm/gos/go1.12.4/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc013b85e60)
        /Users/jzila/.gvm/gos/go1.12.4/src/sync/mutex.go:134 +0x109
sync.(*RWMutex).Lock(0xc013b85e60)
        /Users/jzila/.gvm/gos/go1.12.4/src/sync/rwmutex.go:93 +0x2d
github.com/keybase/client/go/kbfs/libkbfs.(*KBFSOpsStandard).changeHandle(0xc013b85e10, 0x5812680, 0xc004c6ff80, 0x548fbfc, 0x2, 0x1, 0xc00017d380)
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/kbfs_ops.go:1437 +0x4b
github.com/keybase/client/go/kbfs/libkbfs.(*kbfsOpsFavoriteObserver).TlfHandleChange(0xc003d33770, 0x5812680, 0xc004c6ff80, 0xc00017d380)
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/kbfs_ops.go:1591 +0xae
github.com/keybase/client/go/kbfs/libkbfs.(*observerList).tlfHandleChange(0xc003d334a0, 0x5812680, 0xc004c6ff80, 0xc00017d380)
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/observer_list.go:66 +0xc3
github.com/keybase/client/go/kbfs/libkbfs.(*folderBranchOps).Reset.func1()
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/folder_branch_ops.go:8506 +0x45
github.com/keybase/client/go/kbfs/libkbfs.(*folderBranchOps).goTracked.func1(0xc013a85900, 0xc004edb500)
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/folder_branch_ops.go:512 +0x55
created by github.com/keybase/client/go/kbfs/libkbfs.(*folderBranchOps).goTracked
        /Users/jzila/repos/go/src/github.com/keybase/client/go/kbfs/libkbfs/folder_branch_ops.go:510 +0x69
```

Reply:
Which deadlocks with `KBFSOps.Reset`, both of which take `opsLock`.

Reply:
I see. It's not just `opsLock` though, it's because `KBFSOps.Reset()` calls `folderBranchOps.Shutdown()`, which is now waiting on `doneWg`, which now depends on `opsLock`.

This could probably fixed in a more complicated way by `folderBranchOps` returning something that allows `KBFSOps.Reset()` to call the tlf handle change code itself, while still holding the lock.  But if you don't feel like doing that right now and the potential race doesn't seem to hurt things in theory, just leave a comment and I can fix it later.

Reply:
OK I'll do that.